### PR TITLE
Add builder flag to use user-mode only userfaultfd

### LINKED ISSUE #16: Support for unprivileged use in Linux 5.11+
Starting with Linux 5.11, `UFFD_USER_MODE_ONLY` must be passed as a flag to `userfaultfd` to allow for unprivileged use of the system call when the `CAP_SYS_PTRACE` capability isn't present.

With my 5.11.13 kernel, I get an `EPERM` error when attempting to create a `Uffd` object unless I do so from a privileged user.

### PR DESCRIPTION
**Pull request by trinity-1686a**
fix #16

### CODE REVIEW

**Review by pchickey — APPROVED** [review]

**Review by pchickey — CHANGES_REQUESTED** [review]
CI shows that flag isn't available across all Linux versions - can you add the appropriate cfg logic to handle that?

**Comment by trinity-1686a** [comment]
Do you think I should make `UffdBuilder::user_mode_only` always exists, but do nothing when older kernel is used (with doc comment to clarify), or make the function appear only when building with the right flag. I began with the 2nd option as it seemed clearer, but it makes chaining calls like `.close_on_exec(true).non_blocking(true).user_mode_only(true).create` a lot messier

**Comment by olivierlemasle** [comment]
Does that mean the kernel detection should be done by the user of this crate, when deciding if feature `linux5_11` should be enabled? How does a dependent crate (e.g. `wasmtime-runtime`) know if `linux5_11` should be enabled?

I know that this crate previously used linux version detection and switched to features; however with no feature enabled, it worked for all kernels (>= 4.11), without the extra features.

**Comment by trinity-1686a** [comment]
I've changed so at compile time there is no feature-flag. If `UFFD_USER_MODE_ONLY` is not found as part of kernel headers, it's added with the same value as in Linux 5.11, 5.12 and 5.13.
At run-time, the flag is only used when the kernel supports it, so it's always safe to use `user_mode_only(true)` even with an older kernel, and it won't return `EINVAL`

File: tests/manpage_example.rs [hunk 1]
```diff
@@ -1,7 +1,19 @@
 #[test]
 fn run_manpage_example() {
+    #[cfg(feature = "linux5_11")]
+    let args = [
+        "run",
+        "--example",
+        "manpage",
+        "--features",
+        "linux5_11",
```
Comment (pchickey) [inline tests/manpage_example.rs:9]: you can undo the changes in this file - the feature flag no longer exists

File: src/builder.rs [hunk 2]
```diff
@@ -107,6 +123,15 @@ impl UffdBuilder {
...
             flags |= libc::O_NONBLOCK;
         }
+
+        if self.user_mode_only
+            && linux_version::linux_kernel_version()
+                .map(|v| v >= linux_version::Version::new(5, 11, 0))
+                .unwrap_or(true)
+        {
+            flags |= raw::UFFD_USER_MODE_ONLY as i32;
+        }
```
Comment (pchickey) [inline src/builder.rs:133]: I'm not sure I understood your reply in the PR thread - does adding this flag cause an EINVAL on pre-5.11 kernels?

If so please add a comment above this check so we know why we are performing a runtime check

File: linux-version/Cargo.toml [hunk 3]
```diff
@@ -13,4 +13,4 @@ nix = "0.17"
 semver = "0.9"
 
 [build-dependencies]
-bindgen = { version = "0.51", default-features = false }
\ No newline at end of file
+bindgen = { version = "0.57", default-features = false }
```
Comment (pchickey) [inline linux-version/Cargo.toml:16]: was this upgrade required?

File: .github/workflows/rust.yml [hunk 4]
```diff
@@ -31,6 +31,17 @@ jobs:
...
+    # yet for Ubuntu 20.04, so we pull the headers from Ubuntu 21.04.
+    - name: Build (Linux 5.11 support)
+      run:
+        echo 'deb http://cz.archive.ubuntu.com/ubuntu hirsute main' | \
+            sudo tee -a /etc/apt/sources.list &&
+        sudo apt update &&
+        sudo apt install -y linux-headers-5.11.0-16-generic &&
+        export LINUX_HEADERS=/usr/src/linux-headers-5.11.0-16-generic &&
+        cargo build --verbose --features linux5_11
+
```
Comment (pchickey) [inline .github/workflows/rust.yml:44]: remove --features linux5_11 here please

File: userfaultfd-sys/wrapper.h [hunk 5]
```diff
@@ -1 +1,4 @@
 #include <linux/userfaultfd.h>
+#ifndef UFFD_USER_MODE_ONLY
```
Comment (pchickey) [inline userfaultfd-sys/wrapper.h:2]: please add a comment here indicating this definition is not available before linux 5.11, and is provided so that userfaultfd-sys has the same exports on all kernels

File: linux-version/Cargo.toml (hunk 3 above)
Comment (trinity-1686a) [reply linux-version/Cargo.toml:16]: yes, otherwise there was a conflict on native crate `clang-sys`, linux-version transitively depends on v0.28.0, but userfaultfd-sys depends on ^1

File: src/builder.rs [hunk 6]
```diff
@@ -81,6 +84,19 @@ impl UffdBuilder {
...
     }
 
+    /// Enable user-mode only flag for the userfaultfd object.
+    ///
+    /// If set to `false`, the process must have CAP_SYS_PTRACE capability starting with Linux 5.11
+    /// or object creation will fail with EPERM. When set to `true`, the userfaultfd can't be used
+    /// to handle kernel-mode page faults such as when kernel try copying data to userspace.
+    ///
+    /// When used with kernel before version 5.11, this has no effect, the process doesn't need
+    /// CAP_SYS_PTRACE and can handle kernel-mode page fault.
```
Comment (acfoltzer) [inline src/builder.rs:94]:
```suggestion
    /// When used with kernels older than 5.11, this has no effect; the process doesn't need
    /// `CAP_SYS_PTRACE` and can handle kernel-mode page faults.
```

File: src/builder.rs [hunk 7]
```diff
@@ -81,6 +84,19 @@ impl UffdBuilder {
         self
     }
 
+    /// Enable user-mode only flag for the userfaultfd object.
+    ///
+    /// If set to `false`, the process must have CAP_SYS_PTRACE capability starting with Linux 5.11
+    /// or object creation will fail with EPERM. When set to `true`, the userfaultfd can't be used
+    /// to handle kernel-mode page faults such as when kernel try copying data to userspace.
```
Comment (acfoltzer) [inline src/builder.rs:91]:
```suggestion
    /// If set to `false`, the process must have the `CAP_SYS_PTRACE` capability starting with Linux 5.11
    /// or object creation will fail with EPERM. When set to `true`, userfaultfd can't be used
    /// to handle kernel-mode page faults such as when kernel tries copying data to userspace.
```

File: src/builder.rs [hunk 8]
```diff
@@ -60,6 +62,7 @@ impl UffdBuilder {
         UffdBuilder {
             close_on_exec: false,
             non_blocking: false,
+            user_mode_only: false,
```
Comment (acfoltzer) [inline src/builder.rs:65]: I would be inclined to have the default behavior be `user_mode_only: true`, so that only programs which need to be able to handle kernel-mode faults need to set an extra option. Could you make this change, and document its default value in the docstring of this method?

**Review by acfoltzer — CHANGES_REQUESTED** [review]
Thank you! I have a couple documentation tweaks, and a request to change the default behavior of this new flag.

**Comment by olivierlemasle** [comment]
@acfoltzer @pchickey Would you be ok to make a release of the crates after this PR is merged (if PR is ok for you)?

File: src/builder.rs (hunk 2 above)
Comment (ritfed-tabryg) [reply src/builder.rs:135]: I just tested with kernel 4.9 and calling `userfaultfd` with `UFFD_USER_MODE_ONLY` sets errno to EINVAL. I found a more robust way to check was to always try opening the fd with `UFFD_USER_MODE_ONLY` but retry without if it gives EINVAL (here `linux_version` seems to be parsing `uname` which could theoretically be inaccurate if running on a weirdly named kernel).

Comment (trinity-1686a) [reply src/builder.rs:135]: what does `uname -r` report on this machine?

**Review by pchickey — APPROVED** [review]
Sorry we forgot to get back to this one. Github Actions just upgraded the kernel underneath ubuntu-20.04 to 5.11, so we need this fix everywhere downstream to fix CI.
